File: packages/igloo-py/src/igloo_py/element.py
"""Element wrapper and IntoElement protocol."""

from typing import Protocol, Union, Any, TYPE_CHECKING

if TYPE_CHECKING:
    from .types.color import Color

# WitElement is the raw element type from WIT bindings
# After generation, this will be properly typed
WitElement = Any


class IntoElement(Protocol):
    """Protocol for types that can be converted to an Element."""

    def into_element(self) -> "Element":
        """Convert this widget into an Element."""
        ...


class Element:
    """
    Wrapper class for the WIT Element resource.
    Provides a convenient interface for working with UI elements.
    """

    def __init__(self, inner: WitElement) -> None:
        self.inner = inner

    def into_element(self) -> "Element":
        """Returns itself (Element already is an Element)."""
        return self

    def explain(self, color: "Color") -> "Element":
        """
        Debug helper that draws a colored overlay on the element.
        Useful for visualizing element bounds during development.
        """
        # Import here to avoid circular imports
        try:
            from .generated.element import explain  # pyrefly: ignore

            return Element(explain(self.inner, color))
        except ImportError:
            # If bindings not generated yet, return self
            return self


# Type that can be used where an Element is expected
ElementLike = Union[Element, IntoElement]


def to_element(value: ElementLike) -> Element:
    """
    Convert an ElementLike to an Element.
    If already an Element, returns it. Otherwise calls into_element().
    """
    if isinstance(value, Element):
        return value
    return value.into_element()

File: packages/igloo-ts/src/element.ts
import type { Element as WitElement, Color } from 'iced:app/shared@0.1.0';

declare const MSG: unique symbol;

/**
 * Interface for types that can be converted to an Element.
 * Widgets implement this interface to allow automatic conversion.
 *
 * @typeParam Msg - The message type this element (and its subtree) can emit.
 *   A widget that emits no messages is `IntoElement<never>`, which fits any tree.
 */
export interface IntoElement<Msg = never> {
  /**
   * Phantom marker; never present at runtime. Covariant in `Msg` so a widget
   * emitting a subset of the app's messages is assignable where the full union
   * is expected.
   */
  readonly [MSG]?: Msg;

  /** Convert this widget into an Element */
  intoElement(): Element<Msg>;
}

/**
 * Type that can be used where an Element is expected.
 * Either an Element directly, or any widget that implements IntoElement.
 */
export type ElementLike<Msg = never> = Element<Msg> | IntoElement<Msg>;

/**
 * Convert an ElementLike to an Element.
 * If already an Element, returns it. Otherwise calls intoElement().
 */
export function toElement<Msg>(value: ElementLike<Msg>): Element<Msg> {
  if (value instanceof Element) {
    return value;
  }
  return value.intoElement();
}

/**
 * Wrapper class for the WIT Element resource.
 * Provides a convenient interface for working with UI elements.
 *
 * @typeParam Msg - The message type this element's subtree can emit.
 */
export class Element<Msg = never> implements IntoElement<Msg> {
  declare readonly [MSG]?: Msg;

  constructor(public readonly inner: WitElement) {}

  /**
   * Returns itself (Element already is an Element).
   */
  intoElement(): Element<Msg> {
    return this;
  }

  /**
   * Debug helper that draws a colored overlay on the element.
   * Useful for visualizing element bounds during development.
   */
  explain(color: Color): Element<Msg> {
    return new Element(this.inner.explain(color));
  }
}
